### PR TITLE
Fix flakey tests from SSE receiving a 404 response on stop

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/aspnet/SignalR/issues/2034

### CODE REVIEW

File: src/Microsoft.AspNetCore.Http.Connections.Client/Internal/SendUtils.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@ namespace Microsoft.AspNetCore.Http.Connections.Client.Internal
 {
     internal static class SendUtils
     {
-        public static async Task SendMessages(Uri sendUrl, IDuplexPipe application, HttpClient httpClient, ILogger logger)
+        public static async Task SendMessages(Uri sendUrl, IDuplexPipe application, HttpClient httpClient, ILogger logger, CancellationToken cancellationToken)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Http.Connections.Client/Internal/SendUtils.cs]: You can use default here to make this optional and avoid passing in `CancellationToken.None`

File: src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.cs [hunk 2]
```diff
@@ -71,7 +71,7 @@ public LongPollingTransport(HttpClient httpClient, ILoggerFactory loggerFactory)
         {
             // Start sending and polling (ask for binary if the server supports it)
             var receiving = Poll(url, _transportCts.Token);
-            var sending = SendUtils.SendMessages(url, _application, _httpClient, _logger);
+            var sending = SendUtils.SendMessages(url, _application, _httpClient, _logger, CancellationToken.None);
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Http.Connections.Client/Internal/LongPollingTransport.cs]: This wont be needed if you use default

File: src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs [hunk 3]
```diff
@@ -81,14 +81,17 @@ public ServerSentEventsTransport(HttpClient httpClient, ILoggerFactory loggerFac
             _transport = pair.Transport;
             _application = pair.Application;
 
-            Running = ProcessAsync(url, response);
+            CancellationTokenSource inputCts = new CancellationTokenSource();
+            _application.Input.OnWriterCompleted((exception, o) => { ((CancellationTokenSource)o).Cancel(); }, inputCts);
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs]: nit: Please rename `o`

**Comment by JamesNK** [comment]
> checking that the input is completed

How do you do that? I didn't see any flag on the input.

The nice thing about this method and a cancellation token is stop won't need to wait for an in-progress send request to finish.

**Comment by davidfowl** [comment]
The read result that the async loop is using. I dislike this approach because it uses the OnWriterCompleted event to trigger cancellation when you’re already observing the pipe via the async loop. The send request should be short anyways... we can avoid all of this with a simple if statement

**Comment by JamesNK** [comment]
Will the read result change IsCancelled/IsCompleted on `ReadResult` to true AFTER it has been returned? The issue is in the time in-between getting the `ReadResult` and the request being sent and read on the server the connection is closed.

**Comment by davidfowl** [comment]
You would need to do another read. Is that the concern?

**Comment by JamesNK** [comment]
Either the read call would either return a ReadResult with IsComplete true or we'd hang there on the await? That doesn't feel good to me.

I think using OnWriterCompleted, whether it is setting a cancellation token or a simple boolean, is cleaner.

**Comment by JamesNK** [comment]
What's the deal with this kestrel error?

```
Microsoft.AspNetCore.SignalR.Client.FunctionalTests.HubConnectionTests.StreamDoesNotStartIfTokenAlreadyCanceled(protocolName: "messagepack", transportType: WebSockets, path: "/hubT") [FAIL]
  Failed   Microsoft.AspNetCore.SignalR.Client.FunctionalTests.HubConnectionTests.StreamDoesNotStartIfTokenAlreadyCanceled(protocolName: "messagepack", transportType: WebSockets, path: "/hubT")
RUNDOTNET : error Message:  [C:\Users\appveyor\.dotnet\buildtools\korebuild\2.2.0-preview1-17051\KoreBuild.proj]
   System.Exception : 1 error(s) logged.
  SERVER Microsoft.AspNetCore.Server.Kestrel - ApplicationError - Connection id "0HLDRG29NUENM", Request id "0HLDRG29NUENM:00000002": An unhandled exception was thrown by the application.
  ===================
  Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Unexpected end of request content.
     at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.<PumpAsync>d__4.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
     at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
     at System.IO.Pipelines.Pipe.GetReadAsyncResult()
     at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
     at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.<CopyToAsync>d__22.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.<CopyToAsyncInternal>d__23.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher.<ProcessSend>d__16.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher.<ExecuteAsync>d__7.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
     at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
     at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequests>d__186`1.MoveNext()
  ===================
  Stack Trace:
     at Microsoft.AspNetCore.SignalR.Tests.VerifyNoErrorsScope.Dispose()
     at Microsoft.AspNetCore.SignalR.Tests.ServerLogScope.Dispose()
     at Microsoft.AspNetCore.SignalR.Client.FunctionalTests.HubConnectionTests.<StreamDoesNotStartIfTokenAlreadyCanceled>d__13.MoveNext()
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  --- End of stack trace from previous location where exception was thrown ---
     at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
     at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Standard Output Messages:
   | [2018-05-16T15:59:02] TestLifetime Information: Starting test StreamDoesNotStartIfTokenAlreadyCanceled_messagepack_WebSockets_hubT
   | [2018-05-16T15:59:02] ServerLogScope Information: Server log scope started.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:239).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Starting HubConnection.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Server.Kestrel Error: Connection id "0HLDRG29NUENM", Request id "0HLDRG29NUENM:00000002": An unhandled exception was thrown by the application.
   | Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Unexpected end of request content.
   |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.<PumpAsync>d__4.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   |    at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   |    at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   |    at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
   |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.<CopyToAsync>d__22.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.<CopyToAsyncInternal>d__23.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher.<ProcessSend>d__16.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher.<ExecuteAsync>d__7.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   |    at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.<Invoke>d__6.MoveNext()
   | --- End of stack trace from previous location where exception was thrown ---
   |    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   |    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   |    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.<ProcessRequests>d__186`1.MoveNext()
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Starting HttpConnection.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request finished in 43.5549ms 400 
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLDRG29NUENM" stopped.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Establishing connection with server at 'http://127.0.0.1:2854/hubT'.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.LoggingHttpMessageHandler Trace: Sending HTTP request POST 'http://127.0.0.1:2854/hubT/negotiate'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLDRG29NUENN" started.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request starting HTTP/1.1 POST http://127.0.0.1:2854/hubT/negotiate  0
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler Debug: AuthenticationScheme: Bearer was not authenticated.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Routing.RouteBase Debug: Request successfully matched the route with name '(null)' and template '/hubT/negotiate'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionManager Debug: New connection UjzfxsLHF-sywVHB3A2Uvw created.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher Debug: UjzfxsLHF-sywVHB3A2Uvw - Sending negotiation response.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLDRG29NUENN" completed keep alive response.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request finished in 0.7687ms 200 application/json
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Established connection 'UjzfxsLHF-sywVHB3A2Uvw' with the server.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Starting transport 'WebSockets' with Url: http://127.0.0.1:2854/hubT?id=UjzfxsLHF-sywVHB3A2Uvw.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Information: Starting transport. Transfer mode: Binary. Url: 'ws://127.0.0.1:2854/hubT?id=UjzfxsLHF-sywVHB3A2Uvw'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLDRG29NUENO" started.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Hosting.Internal.WebHost Information: Request starting HTTP/1.1 GET http://127.0.0.1:2854/hubT?id=UjzfxsLHF-sywVHB3A2Uvw  
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Authentication.JwtBearer.JwtBearerHandler Debug: AuthenticationScheme: Bearer was not authenticated.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Routing.RouteBase Debug: Request successfully matched the route with name '(null)' and template '/hubT'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.HttpConnectionDispatcher Debug: UjzfxsLHF-sywVHB3A2Uvw - Establishing new connection.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport Debug: UjzfxsLHF-sywVHB3A2Uvw - Socket opened using Sub-Protocol: '(null)'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.SignalR.HubConnectionHandler Debug: UjzfxsLHF-sywVHB3A2Uvw - OnConnectedAsync started.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Transport 'WebSockets' started.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Information: HttpConnection Started.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Information: Using HubProtocol 'messagepack v1'.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending Hub Handshake.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Received message from application. Payload size: 39.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport Trace: UjzfxsLHF-sywVHB3A2Uvw - Message received. Type: Binary, size: 39, EndOfMessage: True.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.SignalR.Internal.DefaultHubProtocolResolver Debug: UjzfxsLHF-sywVHB3A2Uvw - Found protocol implementation for requested protocol: messagepack.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.SignalR.HubConnectionContext Information: UjzfxsLHF-sywVHB3A2Uvw - Completed connection handshake. Using HubProtocol 'messagepack'.
   | [2018-05-16T15:59:02] SERVER Microsoft.AspNetCore.Http.Connections.Internal.Transports.WebSocketsTransport Trace: UjzfxsLHF-sywVHB3A2Uvw - Sending payload: 3 bytes.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Message received. Type: Binary, size: 3, EndOfMessage: True.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Handshake with server complete.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Receive loop starting.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Starting server timeout timer. Duration: 30000.00ms
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Information: HubConnection started.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Releasing Connection Lock in StartAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:281).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in StreamAsChannelCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:364).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.Internal.InvocationRequest.Streaming Trace: Invocation 1 created.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Preparing streaming invocation '1' of 'Stream', with return type 'System.Int32' and 1 argument(s).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Registering Invocation ID '1' for tracking.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Issuing Invocation '1': System.Int32 Stream(System.Int32).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending StreamInvocationMessage message '1'.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending StreamInvocationMessage message '1' completed.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in StreamAsChannelCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:339).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Releasing Connection Lock in StreamAsChannelCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:382).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in StopAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:297).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending Cancellation for Invocation '1'.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending CancelInvocationMessage message '1'.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Sending CancelInvocationMessage message '1' completed.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Releasing Connection Lock in StreamAsChannelCoreAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:356).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.Internal.InvocationRequest.Streaming Trace: Invocation 1 disposed.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Releasing Connection Lock in StopAsyncCore (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:324).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Stopping HubConnection.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Terminating receive loop.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Waiting for the receive loop to terminate.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Received message from application. Payload size: 21.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Waiting on Connection Lock in ReceiveLoop (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:776).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Releasing Connection Lock in ReceiveLoop (/_/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs:785).
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Debug: Disposing HttpConnection.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Information: Transport is stopping.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Send loop stopped.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Transport stopped.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.HttpConnection Information: HttpConnection Disposed.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport Debug: Receive loop stopped.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: Canceling all outstanding invocations.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Removing pending invocation 1.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.Internal.InvocationRequest.Streaming Trace: Invocation 1 disposed.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Trace: Shutting down connection.
   | [2018-05-16T15:59:02] Microsoft.AspNetCore.SignalR.Client.HubConnection Debug: HubConnection stopped.
   | [2018-05-16T15:59:02] ServerLogScope Information: Server log scope stopped.
   | [2018-05-16T15:59:02] TestLifetime Information: Finished test StreamDoesNotStartIfTokenAlreadyCanceled_messagepack_WebSockets_hubT in 0.0229969s
```

File: src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs [hunk 4]
```diff
@@ -81,14 +81,20 @@ public ServerSentEventsTransport(HttpClient httpClient, ILoggerFactory loggerFac
             _transport = pair.Transport;
             _application = pair.Application;
 
-            Running = ProcessAsync(url, response);
+            // Cancellation token will be triggered when the pipe is stopped on the client.
+            // This is to avoid the client throwing from a 404 response caused by the
+            // server stopping the connection while the send message request is in progress.
+            CancellationTokenSource inputCts = new CancellationTokenSource();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs]: var

File: src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs [hunk 5]
```diff
@@ -81,14 +81,20 @@ public ServerSentEventsTransport(HttpClient httpClient, ILoggerFactory loggerFac
...
+            _application.Input.OnWriterCompleted((exception, state) => { ((CancellationTokenSource)state).Cancel(); }, inputCts);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Connections.Client/Internal/ServerSentEventsTransport.cs]: nit

```C#
_application.Input.OnWriterCompleted((exception, state) => ((CancellationTokenSource)state).Cancel(), inputCts);
```

File: test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs [hunk 6]
```diff
@@ -64,5 +55,12 @@ public override IDisposable StartVerifiableLog(out ILoggerFactory loggerFactory,
             var disposable = base.StartVerifiableLog(out loggerFactory, testName, ResolveExpectedErrorsFilter(expectedErrorsFilter));
             return new ServerLogScope(ServerFixture, loggerFactory, disposable);
         }
+
+        public void Dispose()
+        {
+            // Wait for the server to finish processing any in-progress requests
+            // Prevents server logging from a previous tests from showing up in test logs
+            Thread.Sleep(TimeSpan.FromMilliseconds(100));
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs]: This is kinda gross..

Comment (JamesNK) [reply test/Microsoft.AspNetCore.SignalR.Tests.Utils/VerifiableServerLoggedTest.cs]: Yeah. I can't think of a better way to stop server logs from a previous test showing up in the next test. The only surefire way would be to start a new server for each unit test which I imagine would be slower than a small delay after server tests.

**Pull request merged**